### PR TITLE
Refactor istioctl package global variables, and fix some client usages

### PR DESCRIPTION
**Pull request by hanxiaop**
**Please provide a description of this PR:**
This PR also consolidates the usage of the kubeclient.

This is the first step of refactoring, will still need improvements as follow-ups.

### CODE REVIEW

**Comment by hanxiaop** [comment]
https://github.com/istio/istio/issues/38503

**Comment by hanxiaop** [comment]
I thought It's better to separately create things in RunE instead of command initialization, however things for completion like `validPodsNameArgs` are coupled with the kubeclient. I haven't figured out great approaches yet

File: istioctl/pkg/option/option.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+type RootFlags struct {
+	kubeconfig     *string
+	configContext  *string
+	namespace      *string
+	istioNamespace *string
+
+	defaultNamespace string
+}
+
+func NewRootFlags() *RootFlags {
```
Comment (howardjohn) [inline istioctl/pkg/option/option.go:41]: nit: merge AddFlags and NewRootFlags, no need to ever have uninitialized RootFlags?

File: istioctl/pkg/context/context.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+func (c *CLIContext) CLIClient() (kube.CLIClient, error) {
+	return c.CLIClientWithRevision("")
+}
+
+func (c *CLIContext) RestConfig() (*rest.Config, error) {
+	return kube.BuildClientConfig(c.KubeConfig(), c.KubeContext())
+}
+
+func (c *CLIContext) InferPodInfoFromTypedResource(arg string) (pod string, ns string, err error) {
+	if c.factory == nil {
```
Comment (howardjohn) [inline istioctl/pkg/context/context.go:79]: A bit concerned about this lazy initialization. Will try to come back to this with a suggestion

Comment (howardjohn) [reply istioctl/pkg/context/context.go:79]: Also its somewhat confusing that `factory` is one one, but we have a map of clients. Also Factory is basically the same as client - do we need both? (`client.UtilFactory()`)

File: istioctl/pkg/context/context.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	}
+	return c.clients[rev], nil
+}
+
+func (c *CLIContext) CLIClient() (kube.CLIClient, error) {
+	return c.CLIClientWithRevision("")
+}
+
+func (c *CLIContext) RestConfig() (*rest.Config, error) {
+	return kube.BuildClientConfig(c.KubeConfig(), c.KubeContext())
```
Comment (howardjohn) [inline istioctl/pkg/context/context.go:75]: Can be `client.RESTConfig()` from `c.clients`

File: istioctl/pkg/context/context.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+import (
+	"istio.io/istio/istioctl/pkg/option"
+	"istio.io/istio/istioctl/pkg/util/handlers"
+	"istio.io/istio/pkg/kube"
+	"k8s.io/client-go/rest"
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+)
+
+type CLIContext struct {
```
Comment (howardjohn) [inline istioctl/pkg/context/context.go:25]: Naming nit: package name `context` is likely to conflict, so we need to rename it often. Maybe `cli.Context`?

File: istioctl/cmd/workload.go [hunk 5]
```diff
@@ -205,7 +208,7 @@ Configure requires either the WorkloadGroup artifact path or its location on the
 			return nil
 		},
 		RunE: func(cmd *cobra.Command, args []string) error {
-			kubeClient, err := kubeClientWithRevision(opts.Revision)
+			kubeClient, err := kubeClientWithRevision(ctx, opts.Revision)
```
Comment (howardjohn) [inline istioctl/cmd/workload.go:211]: ideally we stop using kubeClientWithRevision. Can be followup though

File: istioctl/cmd/waypoint.go [hunk 6]
```diff
@@ -191,15 +195,15 @@ func waypointCmd() *cobra.Command {
...
 			}
-			ns := defaultNamespace
+			ns := ctx.DefaultNamespace()
 			if allNamespaces {
 				ns = ""
-			} else if namespace != "" {
-				ns = namespace
+			} else if ctx.Namespace() != "" {
+				ns = ctx.Namespace()
 			}
```
Comment (howardjohn) [inline istioctl/cmd/waypoint.go:207]: nit: add NamespaceOrDefault into `ctx` or similar.

Can then replace all of `handlers.HandleNamespace` with it as well

**Review by howardjohn — APPROVED** [review]
Nice work! a few comments, can be followups though

File: istioctl/cmd/waypoint.go (hunk 6 above)
Comment (hanxiaop) [reply istioctl/cmd/waypoint.go:206]: Done.

File: istioctl/cmd/workload.go (hunk 5 above)
Comment (hanxiaop) [reply istioctl/cmd/workload.go:211]: Yes, I was thinking the same. However, some tests are coupled with the func, so I think I'm going to address this in the followup PRs.

File: istioctl/pkg/context/context.go (hunk 4 above)
Comment (hanxiaop) [reply istioctl/pkg/context/context.go]: Thanks for the suggestion, `cli.Context` looks better

File: istioctl/pkg/cli/context.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+// handleNamespace returns the defaultNamespace if the namespace is empty
+func handleNamespace(ns, defaultNamespace string) string {
+	if ns == corev1.NamespaceAll {
+		ns = defaultNamespace
+	}
+	return ns
+}
+
+// TODO(hanxiaop) use interface for context and handle real and fake contexts separately.
+func NewFakeContext(namespace, istioNamespace string) *Context {
```
Comment (hanxiaop) [inline istioctl/pkg/cli/context.go:101]: I'm not sure if having a fake context here is appropriate. However, my thought is that during testing, we could create a mock context similar to `NewFakeClient` to test istioctl commands. Currently, this is used to pass the validate test. Do you have any other ideas, @howardjohn?

If this approach is acceptable, I'll make the context an interface and revise it along with `kubeClientWithRevision` in a follow-up.

**Review by therealmitchconnors — APPROVED** [review]

**Pull request merged**